File: ios/Runner/BeaconSdk/Detection/DetectionEngine.swift
import Foundation

protocol DetectionEngineDelegate: AnyObject {
  func onBeaconRanged(mac: String, locationName: String, rssi: Int, avgRssi: Int, timestamp: Int64, isBackground: Bool, battery: Int?)
  func onBeaconDetected(mac: String, locationName: String, avgRssi: Int, timestamp: Int64, battery: Int?)
  func onBeaconLost(mac: String)
}

class DetectionEngine {
  private let RSSI_BUFFER_SIZE = 5

  private var config: BeaconConfig?
  private weak var delegate: DetectionEngineDelegate?

  private var rssiBuffers: [String: [Int]] = [:]
  private var detectionStartTimes: [String: Int64] = [:]
  private var lastSeenTimes: [String: Int64] = [:]
  private var lastKnownBattery: [String: Int] = [:]

  init(config: BeaconConfig? = nil, delegate: DetectionEngineDelegate? = nil) {
    self.config = config
    self.delegate = delegate
  }
    
  func configure(_ config: BeaconConfig) {
    self.config = config
  }

  func processBeacon(mac: String, locationName: String, rssi: Int, isBackground: Bool, battery: Int?) {
    guard let config = config else {return}

    let now = Int64(Date().timeIntervalSince1970 * 1000)
    lastSeenTimes[mac] = now

    // Not getting battery reading yet
    if let bat = battery {
      lastKnownBattery[mac] = bat
    }
    let currentBattery = lastKnownBattery[mac]

    if rssiBuffers[mac] == nil {rssiBuffers[mac] = [] }
    rssiBuffers[mac]?.append(rssi)

    if let count = rssiBuffers[mac]?.count, count > RSSI_BUFFER_SIZE {
      rssiBuffers[mac]?.removeFirst()
    }

    let buffer = rssiBuffers[mac] ?? []
    let sum = buffer.reduce(0, +)
    let avgRssi = buffer.isEmpty ? rssi:(sum/buffer.count)

    Logger.d("[\(mac)] \(locationName) | RSII: \(rssi)")

    delegate?.onBeaconRanged(
      mac: mac,
      locationName: locationName,
      rssi: rssi,
      avgRssi: avgRssi, 
      timestamp: now,
      isBackground: isBackground,
      battery: currentBattery
    )

    if avgRssi >= config.rssiThreshold {
      handleStrongSignal(mac: mac, locationName: locationName, avgRssi: avgRssi, battery: currentBattery, now: now)
    } else {
      reset(mac: mac)
    }
  }
  

  private func handleStrongSignal(mac: String, locationName: String, avgRssi: Int, battery: Int?, now: Int64) {
    guard let config = config else {return}

    if detectionStartTimes[mac] == nil {
      detectionStartTimes[mac] = now
      Logger.i(" Timer STARTED for \(locationName)")
    }
        
    let startTime = detectionStartTimes[mac]!
    let durationSeconds = (now - startTime) / 1000
    
    if durationSeconds >= Int64(config.timeThreshold) {
      Logger.i(" VALID DETECTION: \(locationName)")
      
      delegate?.onBeaconDetected(
        mac: mac,
        locationName: locationName,
        avgRssi: avgRssi,
        timestamp: now,
        battery: battery
      )

      reset(mac: mac)
    }
  }

  func isWithinShift(start: Double, end: Double, now: Double, earlyIn: Double, lateIn: Double, earlyOut: Double, lateOut: Double) -> Bool {
        
    let earliestCheckIn = start - earlyIn
    let latestCheckIn = start + lateIn
    
    let earliestCheckOut = end - earlyOut
    let latestCheckOut = end + lateOut
    
    let resultIn = (earliestCheckIn...latestCheckIn).contains(now)
    let resultOut = (earliestCheckOut...latestCheckOut).contains(now)
    
    let result = resultIn || resultOut
    
    Logger.i(result ? "Within shift window" : "Outside shift window")
    
    return result
  }

  func checkLostBeacons(timeoutMillis: Int64 = 5000) {
      let now = Int64(Date().timeIntervalSince1970 * 1000)
      
      let lostMacs = lastSeenTimes.filter { (key, lastSeen) in
        return (now - lastSeen) > timeoutMillis
      }
      
      for (mac, _) in lostMacs {
        Logger.i(" Beacon Lost: \(mac) (timed out)")
        reset(mac: mac)
        delegate?.onBeaconLost(mac: mac)
      }
    }
    
  private func reset(mac: String) {
    detectionStartTimes.removeValue(forKey: mac)
    rssiBuffers.removeValue(forKey: mac)
    lastSeenTimes.removeValue(forKey: mac)
  }
}

File: android/beaconsdk/src/main/java/com/xenber/beaconsdk/detection/DetectionEngine.kt
//** Handles RSSI + Time Detection Logic for Target Beacons **/

package com.xenber.beaconsdk.detection

import com.xenber.beaconsdk.BeaconConfig
import com.xenber.beaconsdk.util.Logger
import kotlin.math.abs

class DetectionEngine(
    private val config: BeaconConfig,
    private val listener: DetectionListener
) {

    companion object {
        private const val RSSI_BUFFER_SIZE = 5
        //private const val CHECK_IN_BUFFER = 3600000L 
    }

    private val rssiBuffers = mutableMapOf<String, MutableList<Int>>()      
    private val detectionStartTimes = mutableMapOf<String, Long>()         
    private val lastSeenTimes = mutableMapOf<String, Long>()   
    private var withinShift: Boolean = false
    private val lastKnownBattery = mutableMapOf<String, Int>()         

    fun processBeacon(mac: String, locationName: String, rssi: Int, isBackground: Boolean, battery: Int?) {
        val now = System.currentTimeMillis()
        lastSeenTimes[mac] = now

        val buffer = rssiBuffers.getOrPut(mac) { mutableListOf() }
        buffer.add(rssi)
        if (buffer.size > RSSI_BUFFER_SIZE) buffer.removeAt(0)
        val avgRssi = buffer.average()

        if (battery != null) {
            lastKnownBattery[mac] = battery
        }
        val battery = lastKnownBattery[mac]

        Logger.i("[$mac] $locationName | RSSI: $rssi (avg: ${avgRssi.toInt()})") // | Bat: ${battery ?: "unknown"}%")

        listener.onBeaconRanged(
            mac, locationName, rssi, avgRssi.toInt(), timestamp = now, isBackground, battery
        )

        if (avgRssi >= config.rssiThreshold) {
            handleStrongSignal(mac, locationName, avgRssi, battery)
        } else {
            reset(mac)
        }
    }

    fun isWithinShift(shiftStartTime: Long,shiftEndTime: Long, timestamp: Long, bufferEarlyCheckIn: Long, bufferLateCheckIn: Long, bufferEarlyCheckOut: Long, bufferLateCheckOut: Long): Boolean {

        val earliestCheckInAllowed = shiftStartTime - bufferEarlyCheckIn
        val latestCheckInAllowed = shiftStartTime + bufferLateCheckIn

        val earliestCheckOutAllowed = shiftEndTime - bufferEarlyCheckOut
        val latestCheckOutAllowed = shiftEndTime + bufferLateCheckOut

        val resultIn = timestamp in earliestCheckInAllowed..latestCheckInAllowed
        val resultOut = timestamp in earliestCheckOutAllowed..latestCheckOutAllowed

        val result = resultIn || resultOut

        Logger.i(
            if (result)
                "Within shift window"
            else
                "Outside shift window"
        )

        return result
    }

    private fun handleStrongSignal(mac: String, locationName: String, avgRssi: Double, battery: Int?) {
        val now = System.currentTimeMillis()

        if (!detectionStartTimes.containsKey(mac)) {
            detectionStartTimes[mac] = now
            Logger.i("⏱️ Timer STARTED for $locationName")
        }

        val startTime = detectionStartTimes[mac]!!
        val durationSeconds = (now - startTime) / 1000

        if (durationSeconds >= config.timeThreshold) {
            Logger.i(" VALID DETECTION: $locationName")

            listener.onBeaconDetected(
                mac = mac,
                locationName = locationName,
                avgRssi = avgRssi.toInt(),
                timestamp = now,
                battery = battery
            )

            reset(mac)
        }
    }

    fun checkLostBeacons(timeoutMillis: Long = 5000) {
        val now = System.currentTimeMillis()
        val lostMacs = lastSeenTimes.filter { now - it.value > timeoutMillis }.keys

        lostMacs.forEach { mac ->
            Logger.i("🔌 Beacon Lost: $mac (timed out after ${timeoutMillis}ms)")
            reset(mac)
            listener.onBeaconLost(mac)
        }
    }

    private fun reset(mac: String) {
        detectionStartTimes.remove(mac)
        rssiBuffers.remove(mac)
        lastSeenTimes.remove(mac)
    }

    interface DetectionListener {
        fun onBeaconRanged(
            mac: String,
            locationName: String,
            rssi: Int,
            avgRssi: Int,
            timestamp: Long,
            isBackground: Boolean,
            battery: Int?
        )

        fun onBeaconDetected(
            mac: String,
            locationName: String,
            avgRssi: Int,
            timestamp: Long,
            battery: Int?
        )

        fun onBeaconLost(mac: String)
    }
}